DRILL-7445: Create batch copier based on result set framework

paul-rogers opened the pull request:
The result set framework now provides both a reader and writer.
This PR provides a copier that copies batches using this
framework. Such a copier can:

- Copy selected records
- Copy all records, such as for an SV2 or SV4

The copier uses the result set loader to create uniformly-sized
output batches from input batches of any size. It does this
by merging or splitting input batches as needed.

Since the result set reader handles both SV2 and SV4s, the
copier can filter or reorder rows based on the SV associated
with the input batch.

This version assumes single stream of input batches, and handles
any schema changes in that input by creating output batches
that track the input schema. This would be used in, say, the
selection vector remover. A different design is needed for merging
such as in the merging receiver.

Adds a "copy" method to the column writers. Copy is implemented
by doing a direct memory copy from source to destination vectors.

A unit test verifies functionality for various use cases
and data types.

ihuzenko commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/resultSet/impl/ResultSetCopierImpl.java:
```diff
@@ -0,0 +1,321 @@
+    project();
+  }
+
+  private final void project() {
```
```suggestion
  private void copyInputRow() {
```

paul-rogers replied:
Renamed to `copyColumns()` (which is what "project" means.) The `final` has meaning for a method: it means that there will never be overrides and so the JVM can optimize calls, which is helpful given that this is called once per rows.

ihuzenko replied:
Thanks for explaining the reasons, but private methods already can't be overridden. Though this was actually minor suggestion.

ihuzenko commented on exec/java-exec/src/main/java/org/apache/drill/exec/record/selection/SelectionVector2Builder.java:
```diff
@@ -0,0 +1,47 @@
+import org.apache.drill.exec.memory.BufferAllocator;
+import org.apache.drill.exec.record.VectorContainer;
+
+public class SelectionVector2Builder {
```
Seems like the class is only intended for test purposes. If so, please move to test sources.

paul-rogers replied:
That's where it is used in this PR. However, when doing the batch fixes, I found that we repeated the same logic (often incorrectly) in multiple places. The intent is to use this class in those cases as well. By putting it in this PR we can review it without also having to understand the operators that will eventually be changed.

ihuzenko commented on exec/vector/src/main/codegen/templates/ColumnAccessors.java:
```diff
@@ -592,6 +592,32 @@ public final void setDefaultValue(final Object value) {
+      setBytes(bytes, bytes.length);
+    <#else>
+      set${label}(source.get${label}());
+    </#if> -->
```
please remove commented lines

ihuzenko commented on exec/vector/src/main/java/org/apache/drill/exec/vector/accessor/convert/AbstractWriteConverter.java:
```diff
@@ -143,4 +144,9 @@ public void setTime(LocalTime value) {
+
+  @Override
+  public void copy(ColumnReader from) {
+    throw new UnsupportedOperationException("Cannot copy values through a type converter");
```
If type converters shouldn't support copy then maybe makes sense to mark the method as final here.

ihuzenko commented on exec/vector/src/main/java/org/apache/drill/exec/vector/accessor/writer/UnionWriterImpl.java:
```diff
@@ -332,6 +334,14 @@ public int writeIndex() {
 
+  @Override
+  public void copy(ColumnReader from) {
+    if (! from.isNull()) {
```
```suggestion
    if (!from.isNull()) {
```

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/resultSet/impl/ResultSetCopierImpl.java:
```diff
@@ -0,0 +1,321 @@
+      OptionBuilder outputOptions) {
+    this.allocator = allocator;
+    resultSetReader = new ResultSetReaderImpl(inputBatch);
+    writerOptions = outputOptions;
```
Seems overly clever to save one line.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/resultSet/impl/ResultSetCopierImpl.java:
```diff
@@ -0,0 +1,321 @@
+    resultSetReader = new ResultSetReaderImpl(inputBatch);
+    writerOptions = outputOptions;
+    writerOptions.setVectorCache(new ResultVectorCacheImpl(allocator));
+    state = State.START;
```
See above about use of `this.` prefix.

paul-rogers commented:
@ihuzenko, thanks for your detailed review. Addressed most of your comments, added explanations for the others.

To provide a bit of context: this is a first step toward additional cleanup in the operators ("record batches") based on issues I've been finding as I work though the batch count, row count, vector value count issues. Did this change as its own PR to make it easier to review. A subsequent PR will use these changes to modify selected operators.

ihuzenko commented on common/src/main/java/org/apache/drill/common/types/Types.java:
```diff
@@ -806,16 +806,20 @@ public static boolean isSortable(MinorType type) {
     return typeBuilder;
   }
 
+  public static boolean isSameType(MajorType type1, MajorType type2) {
```
Yes I got the idea, and this suggestion is part of my personal beliefs. I like to hide everything until it's required to be used elsewhere. Since it's minor comment, I'm ok with leaving it as is.
